test/stream-for-each.cc

```diff
@@ -0,0 +1,271 @@
+        | Collect<std::vector<int>>()
+              .stop([](auto& data, auto& k) {
+                k.Start(std::move(data));
+              });
+  };
+
+  auto [future, k] = Terminate(e());
+
+  Interrupt interrupt;
+
+  k.Register(interrupt);
+
+  k.Start();
+
+  interrupt.Trigger();
+
+  EventLoop::Default().Run();
+
+  auto result = future.get();
+
+  CHECK_EQ(result.size(), 0);
+}
+
+TEST(StreamForEach, InterruptReturn) {
+  std::atomic<bool> waiting = false;
+  std::atomic<bool> interrupted = false;
+
+  auto e = [&]() {
+    return Iterate(std::vector<int>(1000))
+        | StreamForEach([&](int x) {
+             return Stream<int>()
+                 .start([&](auto& k) {
+                   waiting.store(true);
+                 })
+                 .next([](auto& k) {
+                   k.Ended();
+                 })
+                 .interrupt([&](auto& k) {
+                   interrupted.store(true);
+                   k.Stop();
+                 });
+           })
+        | Collect<std::vector<int>>();
+  };
+
+  auto [future, k] = Terminate(e());
+
+  Interrupt interrupt;
+
+  k.Register(interrupt);
+
+  ASSERT_FALSE(waiting.load());
+  ASSERT_FALSE(interrupted.load());
+
+  k.Start();
+
+  ASSERT_TRUE(waiting.load());
+  ASSERT_FALSE(interrupted.load());
+
+  interrupt.Trigger();
+
+  EXPECT_THROW(future.get(), stout::eventuals::StoppedException);
+
+  ASSERT_TRUE(interrupted.load());
+}
+
+// Iterate with init_list
+TEST(StreamForEach, Let) {
+  auto s = []() {
+    return Iterate(std::vector<int>{1, 2})
+        | StreamForEach(Let([](int& x) {
+             return Iterate(std::vector<int>{1, 2})
+                 | StreamForEach(Let([&x](int& y) {
+                      return Iterate(std::vector<int>{x, y});
+                    }));
+           }))
+        | Collect<std::vector<int>>();
+  };
+
+  EXPECT_THAT(*s(), ElementsAre(1, 1, 1, 2, 2, 1, 2, 2));
+}
+
+TEST(StreamForEach, TwoLevelLoop) {
+  auto s = []() {
+    return Range(2)
+        | StreamForEach([](int x) { return Range(2); })
+        | Collect<std::vector<int>>();
+  };
+
+  EXPECT_THAT(*s(), ElementsAre(0, 1, 0, 1));
+}
+
+TEST(StreamForEach, TwoIndexesSum) {
+  auto s = []() {
+    return Range(3)
+        | StreamForEach([](int x) {
+             return Stream<int>()
+                 .next([container = std::vector<int>({1, 2}),
+                        i = 0u,
+                        x](auto& k) mutable {
+                   if (i != container.size()) {
+                     k.Emit(container[i++] + x);
+                   } else {
+                     k.Ended();
+                   }
+                 })
+                 .done([](auto& k) {
+                   k.Ended();
+                 });
+           })
+        | Collect<std::vector<int>>();
+  };
+
+  EXPECT_THAT(*s(), ElementsAre(1, 2, 2, 3, 3, 4));
+}
+
+TEST(StreamForEach, TwoIndexesSumMap) {
+  auto s = []() {
+    return Range(3)
+        | StreamForEach([](int x) { return Range(1, 3)
```

Comment:
Let's always try and wrap functions like this on a newline, it's easier to read the return and continuation that way. Here and below, thanks!